lib/mutgen.c

```diff
@@ -434,28 +714,320 @@ mutgen_generate(mutgen_t *self, tsk_table_collection_t *tables, int flags)
+    double u = gsl_ran_flat(rng, 0.0, 1.0);
+    int j = 0;
+
+    while (u > probs[j]) {
+        u -= probs[j];
+        j++;
+        assert(j < (int) self->num_alleles);
+    }
+    return j;
+}
+
+static int
+mutgen_set_ancestral_state(mutgen_t *self, site_t *site, int allele)
+{
+    int ret = 0;
+    size_t len = strlen(self->model->alleles[allele]);
+    char *buff = tsk_blkalloc_get(&self->allocator, len);
+
+    if (buff == NULL) {
+        ret = MSP_ERR_NO_MEMORY;
+        goto out;
+    }
+    memcpy(buff, self->model->alleles[allele], len);
+    site->ancestral_state = buff;
+    site->ancestral_state_length = (tsk_size_t) len;
+out:
+    return ret;
+}
+
+static int
+mutgen_set_derived_state(mutgen_t *self, mutation_t *mut, int allele)
+{
+    int ret = 0;
+    size_t len = strlen(self->model->alleles[allele]);
+    char *buff = tsk_blkalloc_get(&self->allocator, len);
+
+    if (buff == NULL) {
+        ret = MSP_ERR_NO_MEMORY;
+        goto out;
+    }
+    memcpy(buff, self->model->alleles[allele], len);
+    mut->derived_state = buff;
+    mut->derived_state_length = (tsk_size_t) len;
+out:
+    return ret;
+}
+
+static int MSP_WARN_UNUSED
+mutgen_choose_alleles(mutgen_t *self, tsk_id_t *parent, mutation_t **bottom_mutation,
+        tsk_size_t num_nodes, site_t *site)
+{
+    int ret = 0;
+    int ai, pa, da;
+    mutation_t *mut, *parent_mut;
+    tsk_id_t u;
+
+    ret = sort_mutations(site);
+    if (ret != 0) {
+        goto out;
+    }
+
+    if (site->new) {
+        ai = mutation_model_pick_allele(self->model, self->rng);
+        ret = mutgen_set_ancestral_state(self, site, ai);
+        if (ret != 0) {
+            goto out;
+        }
+    } else {
+        /* we don't error here if the allele is not found (and so ai < 0)
+         * because it's ok if we don't try to add a mutation to the site,
+         * e.g., if it is kept and we're doing infinite sites, or if we're
+         * mutating a different part of the genome  */
+        ai = mutation_model_allele_index(self->model,
+                site->ancestral_state, site->ancestral_state_length);
+    }
+
+    /* Create a mapping from mutations to nodes in bottom_mutation. If we see
+     * more than one mutation at a node, the previously seen one must be the
+     * parent of the current one since we assume they are in order. */
+    for (mut = site->mutations; mut != NULL; mut = mut->next) {
+        u = mut->node;
+        assert((tsk_size_t) u < num_nodes);
+        while (u != TSK_NULL && bottom_mutation[u] == NULL) {
+            u = parent[u];
+        }
+        if (u == TSK_NULL) {
+            pa = ai;
+            assert(mut->parent == NULL);
+        } else {
+            parent_mut = bottom_mutation[u];
+            mut->parent = parent_mut;
+            if (mut->time > parent_mut->time || (parent_mut->new && !mut->new)) {
+                ret = MSP_ERR_MUTATION_GENERATION_OUT_OF_ORDER;
+                goto out;
+            }
```

Comment:
This is only bit I'm still uneasy with. I tried writing a test case to provoke it but couldn't. How can this happen?

Reply:
Gee, I'm provoking it [here](https://github.com/tskit-dev/msprime/pull/946/files#diff-f8fd4eeeb7bf8fea997396a5125091e4R4320) and [here](https://github.com/tskit-dev/msprime/pull/946/files#diff-09dc053e0ddd2b7c406d68a525d4a348R444), I think?

Reply:
Oh wait, maybe that's the other time that comes up... lemme see...

Reply:
Well, [codecov](https://codecov.io/gh/tskit-dev/msprime/compare/1e267b43d9bed27a8d8d376666ce5d9061706ddb...52288dd3374cf7e5dc3523cc239260ab487fe0e2/src/lib/mutgen.c?before=lib/mutgen.c#L870) says we hit this, but not the [other place](https://codecov.io/gh/tskit-dev/msprime/compare/1e267b43d9bed27a8d8d376666ce5d9061706ddb...52288dd3374cf7e5dc3523cc239260ab487fe0e2/src/lib/mutgen.c?before=lib/mutgen.c#L647) this error might be thrown.  That's because the second one should really be an assert; if the algorithm's right it should not happen at that point; I'll change that.